Make noexcept for AssertNoThrow more explicit

masterleinad opened the pull request:
`AssertNoThrow` used to call `issue_error` which may throw. In particular, Coverity complains about this.
Also make sure to not indirectly call `MPIThrow` in `internals::abort` through `Utilities::MPI::n_mpi_processes`

drwells commented:
Relates to #4704.

drwells commented on include/deal.II/base/exceptions.h:
```diff
@@ -298,13 +299,34 @@ namespace deal_II_exceptions
+      e.set_fields (file, line, function, cond, exc_name);
+      // We are not allowed to throw, and not allowed to abort.
+      // Just print the exception name to deallog and continue normally:
+      std::cerr << "Exception: " << e.get_exc_name() << std::endl;
```
Could we add a call to `what()` here to print out some more details?

drwells commented:
This looks like the right way to do things.

bangerth commented on include/deal.II/base/exceptions.h:
```diff
@@ -243,7 +244,7 @@ namespace deal_II_exceptions
      * <tt>throw()</tt>).
      */
-    void abort (const ExceptionBase &exc, bool nothrow = false);
+    void abort (const ExceptionBase &exc);
```
you'll have to update the documentation

bangerth commented on include/deal.II/base/exceptions.h:
```diff
@@ -298,13 +299,34 @@ namespace deal_II_exceptions
+     * @ref ExceptionBase
+     */
+    template <class ExceptionType>
+    void issue_error_nothrow (ExceptionHandling,
```
you could make the last argument of type `ExceptionBase`. In that case, you don't need the template, can move the implementation into the .cc file, and remove the `#include <iostream>` again.

masterleinad commented:
@bangerth @drwells I addressed your comments.

drwells commented:
Nice work. Would you also please add a changelog entry (I suppose that, legally, this is incompatible)?

It would also be nice to add a simple test to make sure that we print everything we think we print to `cerr` and then return exit status `0` when we execute an `AssertNothrow`.

masterleinad commented:
@tjhei I extended the explanation a bit. Is this clearer?

@drwells Rethinking this I decided to not change the behavior of `AssertNothrow` and updated the PR in this regard. Hence, I don't think that a changelog entry makes sense. I added a test as well.

tjhei commented:
> @tjhei I extended the explanation a bit. Is this clearer?

:+1:

tjhei commented on source/base/exceptions.cc:
```diff
@@ -399,68 +399,87 @@ namespace StandardExceptions
+    MPI_Initialized(&is_initialized);
+    if (is_initialized)
+      {
+        // do the same as in Utilities::MPI::n_mpi_processes() here,
```
"but without error checking to not throw again"

masterleinad replied:
I copied your line below.

masterleinad commented:
Now the test output makes sense as well.

masterleinad commented:
/run-tests

masterleinad commented:
> lac/tensor_product_matrix_03.debug

is the only failing test and unrelated.

drwells commented:
This looks very nice. I'll merge in a day or so unless someone else wants to comment.
